streams/src/test/java/org/apache/kafka/streams/integration/PositionCheckpointIntegrationTest.java

```diff
@@ -0,0 +1,777 @@
+
+    public void reboot() {
+        afterTest(false);
+        beforeTest(false);
+    }
+
+    @Test
+    public void verifyStore() throws Exception {
+        if (storeToTest.keyValue()) {
+            shouldCheckpointKeyValueStore(2);
+        }
+
+        if (storeToTest.isWindowed()) {
+            shouldHandleWindowKeyQueries();
+        }
+
+        if (storeToTest.isSession()) {
+            shouldHandleSessionKeyQueries();
+        }
+    }
+    public <V> void shouldCheckpointKeyValueStore(
+            final Integer key) throws Exception {
+        shouldHandleKeyQuery(key);
+        reboot();
+        shouldHandleKeyQuery(key);
+    }
+
+    private <T> void shouldHandleWindowKeyQueries() {
+        shouldHandleWindowKeyQuery(
+            2,
+            Instant.ofEpochMilli(WINDOW_START),
+            Instant.ofEpochMilli(WINDOW_START)
+        );
+        reboot();
+        shouldHandleWindowKeyQuery(
+                2,
+                Instant.ofEpochMilli(WINDOW_START),
+                Instant.ofEpochMilli(WINDOW_START)
+        );
+    }
+
+    private <T> void shouldHandleSessionKeyQueries() {
+        shouldHandleSessionRangeQuery(
+            2,
+            mkSet(2)
+        );
+        reboot();
+        shouldHandleSessionRangeQuery(
+                2,
+                mkSet(2)
+        );
+    }
+
+    public <V> void shouldHandleKeyQuery(
+            final Integer key) {
+
+        final KeyQuery<Integer, V> query = KeyQuery.withKey(key);
+        final StateQueryRequest<V> request =
+                inStore(STORE_NAME)
+                        .withQuery(query)
+                        .withPartitions(mkSet(0, 1))
+                        .withPositionBound(PositionBound.at(INPUT_POSITION));
+
+        final StateQueryResult<V> result =
+                IntegrationTestUtils.iqv2WaitForResult(kafkaStreams, request);
+        final QueryResult<V> queryResult = result.getOnlyPartitionResult();
+        final boolean failure = queryResult.isFailure();
+        if (failure) {
+            throw new AssertionError(queryResult.toString());
+        }
+        assertThat(queryResult.isSuccess(), is(true));
+    }
+
+    public <V> void shouldHandleWindowKeyQuery(
+        final Integer key,
+        final Instant timeFrom,
+        final Instant timeTo) {
+
+        final WindowKeyQuery<Integer, V> query = WindowKeyQuery.withKeyAndWindowStartRange(
+            key,
+            timeFrom,
+            timeTo
+        );
+
+        final StateQueryRequest<WindowStoreIterator<V>> request =
+            inStore(STORE_NAME)
+                .withQuery(query)
+                .withPartitions(mkSet(0, 1))
+                .withPositionBound(PositionBound.at(INPUT_POSITION));
+
+        final StateQueryResult<WindowStoreIterator<V>> result = IntegrationTestUtils.iqv2WaitForResult(kafkaStreams, request);
+
+        if (result.getGlobalResult() != null) {
+            fail("global tables aren't implemented");
+        }
+    }
```

Comment:
In an earlier version I had an extra assertion here: assertThat(result.getPosition(), is(INPUT_POSITION)), but turned out that result.getPosition never matched.

Reply:
huh, that's weird. It seems like it should, especially since that's also the bound, so the query shouldn't succeed at all unless it's at that position.

Reply:
yea it's weird. digging into it a bit further I see that despite the result being correct the underlying partition results have a "null" position, therefore the merging in result.getPosition() leads to the empty Position. Again that assertion also fails when being added to the IQv2IntegrationTest.